### ISSUE TITLE
Should custom elements be adoptable if so, how?

### ISSUE THREAD

**Issue by rniwa** [WICG/webcomponents#512]
We haven't figured out how a custom element may be adopted or it should support adaptation at all.

It seems inevitable to support adaptation but there was a big controversy the last time we discussed this, which was whether the prototype of an adopted custom element should be changed or not.

WebKit and Blink preserve the original document's prototype whereas Firefox and Edge(?) change it.

**Comment by rniwa** [comment]
@bzbarsky @annevk @travisleithead @domenic

**Comment by domenic** [comment]
Thanks for raising this. Good to have the discussion. /cc @dominiccooney 

> It seems inevitable to support adaptation but there was a big controversy the last time we discussed this, which was whether the prototype of an adopted custom element should be changed or not.

If I understand the current situation correctly, the DOM spec aligns with WebKit/Blink behavior. There seems to be an irreconcilable difference of opinion as to which behavior should be done and since it hasn't caused any compat issues everyone seems to be in an uneasy stalemate in that regard. (This is all unrelated to custom elements.)

For custom elements my thought was that we would not take an opinion on this question. If we don't do so, then adoption kind of falls into place automatically:
- Adopting should downgrade an element's status to "undefined" (so :defined doesn't match it and no callbacks fire anymore)
  - In WebKit/Blink, the prototype will stay the same as in the old window, as in the current DOM spec for adoption.
  - In Firefox/Edge?, the prototype will probably be changed to something like HTMLElement? unsure; that behavior is unspecced.
- Once adopted, inserting into a new document will try perform an upgrade
  - If there is no element defined with that name in the new document's window's registry, then things will be left as they were before (with different behavior in different browsers)
  - If there _is_ such a definition, then this automatically causes proto-munging, as all upgrades always do. The end result will be the same in all browsers.

**Comment by rniwa** [comment]
> For custom elements my thought was that we would not take an opinion on this question.

I don't think we can decide what custom elements do without agreeing on what regular elements should do.

> Adopting should downgrade an element's status to "undefined" (so :defined doesn't match it and no callbacks fire anymore)

That doesn't make any sense to me.  An element doesn't suddenly get un-created just because it got inserted into a new document.  e.g. an input element that gets adopted into a new document preserves the dirty value ([see demo](https://jsfiddle.net/k1horvft/1/)).

Also, calling different constructors twice on the same element seems like an abuse.

**Comment by annevk** [comment]
See https://www.w3.org/Bugs/Public/show_bug.cgi?id=20567#c69 onward for the conclusions from last time. That was mainly blocked on custom elements getting defined.

I don't think Mozilla's position on this has changed. We'd still favor changing prototypes for built-ins and giving custom elements the opportunity to do so (or do it automatically?) so as to reduce the risk of leaking globals.

Not sure how keen @bholley is on doing this debate again, but paging him just the same.

**Comment by rniwa** [comment]
Key problem here is that when a custom element is adopted from one document to another document, there is no guarantee that the new owner document's global object defines the same custom element.

So changing the prototype internal slot of an adopted element raises a question as to which value we're changing it for custom elements.  And it doesn't seem desirable for a custom element to be completely broken as soon as it's adopted into a document.  Furthermore, doing so could result in an unexpected behavior of an element completely changing its interface and behavior if the new owner document's custom element definition didn't match that of the old owner document.

**Comment by annevk** [comment]
The idea is that we'd provide a callback. For built-in elements the implementation of that callback would be to change the prototype. For custom elements you need to implement that behavior yourself. I agree that we can't really do better than that.

**Comment by dominiccooney** [comment]
I'm glad you brought this up @rniwa. We should do something for adoption. In the spec today, the current document's definitions are used for finding callbacks, which means element that's moved between documents will pick up callbacks from a definition it never ran the constructor for.

> [@domenic's proposal] doesn't make any sense to me.  An element doesn't suddenly get un-created just because it got inserted into a new document.

Being "un-created" would not happen to the JavaScript object either; the same object would continue to exist. This is similar to the case where an existing element is upgraded: the element existed before it was upgraded and it is the same object after it has been upgraded. It has just been mutated.

> calling different constructors twice on the same element seems like an abuse.

It does seem odd. On the other hand @domenic 's proposal does make the following case consistent: Say you have document A and document B. Document B has a definition for my-button.

I think with the spec today, if you create a my-button in document A and move it to document B, it may or may not get upgraded. It depends on whether document A had a definition for my-button or not.

With Domenic's proposal, it's simple and consistent: my-button gets the definition in document B, regardless of what happened to it in the past.

> The idea is that we'd provide a callback.

@annevk I suppose it is the definition in the original document which gets the callback? That's the only definition that is guaranteed to exist. I think this makes sense.

One question I have is running callbacks when the window is gone. If an element from document A is squirreled away somewhere, then A's window is closed/navigated/etc. and later the element is adopted into document B. The document change happens at that point, but it doesn't seem right to be running script from a closed window. It's true script that was already running continues to run after the window is closed, but I'm not sure how readily Blink would start new invocations when there's no window.

This was one drawback of the original custom elements spec. Even if you control all of your documents and can register the same definition in all of them, once you close a window/iframe/etc. all of the elements which came from that context stop firing callbacks and are dead in the water. The built-in elements don't work that way.

Domenic's proposal does not suffer from these drawbacks because it only relies on the definitions in the new document.

**Comment by rniwa** [comment]
> One question I have is running callbacks when the window is gone. If an element from document A is squirreled away somewhere, then A's window is closed/navigated/etc. and later the element is adopted into document B. The document change happens at that point, but it doesn't seem right to be running script from a closed window. It's true script that was already running continues to run after the window is closed, but I'm not sure how readily Blink would start new invocations when there's no window.

How is that different from grabbing a JS closure function out of an iframe that then gets removed?  The closure keeps around the global object of the removed iframe (and its document) in that case as well.

> This was one drawback of the original custom elements spec. Even if you control all of your documents and can register the same definition in all of them, once you close a window/iframe/etc. all of the elements which came from that context stop firing callbacks and are dead in the water. The built-in elements don't work that way.

That just seems like a bug in the engine.  Do you also prevent authors from calling a function that comes from a removed iframe/window?

> I think with the spec today, if you create a my-button in document A and move it to document B, it may or may not get upgraded. It depends on whether document A had a definition for my-button or not.
> 
> With Domenic's proposal, it's simple and consistent: my-button gets the definition in document B, regardless of what happened to it in the past.

I think this is really a problem of perception.  If I created an empty iframe with no scripts, I can adopt an input element into it and the element continues to function.  With @domenic's proposal, I can't do the same for custom elements.  In effect, I would have to include all the custom elements' definitions in every single document into which I may adopt my custom element.  That seems like an awful lot of overhead for a simple use case of wanting to create an editable iframe and put some widgets into it for example.

**Comment by bzbarsky** [comment]
Trying to respond to the various bits, but not necessarily in order:

> Do you also prevent authors from calling a function that comes from a removed iframe/window?

Edge does this sort of thing in a lot of cases.  For example, it _will_ prevent authors from calling a function from a navigated-away-from iframe/window.  Not sure about the removed case, but when Gecko tried to prevent those being called as callbacks (e.g. event listeners from a removed iframe set on elements that are in a non-removed iframe) I think we ran into web compat issues.  I would be quite interested in thorough testing of UA behavior around this, though it's moderately off-topic for this issue.

> WebKit and Blink preserve the original document's prototype whereas Firefox and Edge(?) change it.

Did WebKit change the behavior since I last checked?  It used to have gc-dependent behavior here; see https://lists.w3.org/Archives/Public/www-dom/2012OctDec/0149.html

**Comment by domenic** [comment]
@rniwa

> I don't think we can decide what custom elements do without agreeing on what regular elements should do.

I disagree, but we can set that aside for now.

> That doesn't make any sense to me. An element doesn't suddenly get un-created just because it got inserted into a new document. e.g. an input element that gets adopted into a new document preserves the dirty value (see demo).
> 
> Also, calling different constructors twice on the same element seems like an abuse.

These are strong points and I am unsure what is best. Constructors are generally constructed with the invariant that they only execute on uninitialized objects. We're already stretching that greatly in the upgrade process, and running two different constructors on the same object pushes us pretty far, perhaps past the breaking point.

With that in mind, if we want to avoid running constructors twice, I see two options:
- Custom elements are modified carry around a pointer to their element definition. Thus, after adoption, mutations of the element will continue working, using the old element definition from the old window. This involves all the potential issues with running script in a no-longer-existing window, of course.
- We do nothing. Thus, after adoption, mutations of the element might coincidentally trigger the "wrong" callbacks, since the element's node document's window's custom element registry is used for lookups, and that might have an element definition with the same name as the element. Or it might not, in which case no callbacks fire.
  - In some ways this is not as bad as it sounds, if you think of an element as being defined by its tag name
  - But, the fact that the constructor from the new window will not have gotten a chance to run, even though attributeChanged etc. will do so, seems pretty darn bad. Maybe it could be salvaged with an adoption callback... but I'd almost rather just run the constructor...

These two options are not great either.

However, we really need to pick _some_ semantics. We can't just say "we don't support adoption"; we need a constructive counterproposal for _how_ we're going to not support adoption. I guess one would be throwing in any adoption that tries to adopt a custom element...

**Comment by rniwa** [comment]
We may need to throw in the case of trying to adopt a custom element until the general problem of adopting node into another document.  I think trying to come up with any solution without agreeing on the generic case can result in an undesirable inconsistencies between builtin elements and custom elements.

**Comment by domenic** [comment]
I think there will always be inconsistencies since built-in elements have "the same" definition in all windows. It can't really be avoided.

**Comment by rniwa** [comment]
> I think there will always be inconsistencies since built-in elements have "the same" definition in all windows. It can't really be avoided.

I disagree.  If we picked the behavior WebKit/Blink uses which is to keeping the original prototype and pull the definition from the old window, then custom elements will continue to function and appears to behave identically to builtin elements.

But if we can't agree on changing engine's behaviors for this issue, then we should just throw on adaptation so that we wouldn't be stuck with a forward incompatible behavior.

**Comment by domenic** [comment]
> I disagree. If we picked the behavior WebKit/Blink uses which is to keeping the original prototype and pull the definition from the old window, then custom elements will continue to function and appears to behave identically to builtin elements.

Do you think it would be acceptable to spec this behavior for custom elements (just like it is currently specced for built-in elements), despite Firefox not following the spec for built-in elements? Or do you think it's only OK to spec a useful behavior for custom elements if Firefox first changes its behavior for built-in elements to follow the current spec?

**Comment by rniwa** [comment]
> > I disagree. If we picked the behavior WebKit/Blink uses which is to keeping the original prototype and pull the definition from the old window, then custom elements will continue to function and appears to behave identically to builtin elements.
> 
> Do you think it would be acceptable to spec this behavior for custom elements (just like it is currently specced for built-in elements), despite Firefox not following the spec for built-in elements? Or do you think it's only OK to spec a useful behavior for custom elements if Firefox first changes its behavior for built-in elements to follow the current spec?

I don't think it's acceptable to be specifying this behavior for custom elements without reaching a consensus on what happens to adopted builtin elements unless we delegate the entire work to authors e.g. by simply calling `adoptedCallback` and not doing anything in UA.

However, I would be surprised if @bzbarsky and others at Mozilla are okay with that kind of behavior since that would mean that all custom elements adopted into another document would leak the original document if they're not okay with leaking the original owner document of builtin elements.

**Comment by dominiccooney** [comment]
I guess Blink probably can run the callbacks of a detached window in the new document. It will be a bit tough on authors because they have to write something like `this.document.window.X` where you would normally write `X`, if X depends on things like having an active browsing context&mdash;which so much stuff does.

This would extend to libraries the callbacks used; they have the disadvantage of not having access to 'this' to get the new window, unless it is passed to them. This might make a lot of libraries useless in adoptable custom elements.

**Comment by domenic** [comment]
Let's try to drive a resolution on this. I see a few possibilities.

### Proposals

#### Throw when adopting

This would make adopting a custom element throw.

The spec remains as-is for uncustomized elements (no proto swizzling), and some browsers presumably continue to not comply with it.

The main downside of this is that it's not clear whether it's safe to introduce throwing into adoption. I believe some browsers perform adoption frequently (e.g. while printing?) and won't be expecting failure.

#### Stay custom with same definition

Adopting a custom element leaves it with a custom element state of "custom". It carries with it its original custom element definition and prototype.

The spec remains as-is for uncustomized elements (no proto swizzling), and some browsers presumably continue to not comply with it.

In the future we could perhaps introduce an adoptedCallback. This is a good idea anyway to allow custom elements to have "adoption steps". (Although only one built-in element, `<template>`, uses these.) It also allows the non-compliant browsers to rationalize their behavior by saying that the default adoptedCallback equivalent for built-in elements does proto-swizzling, even if the spec does not mention that.

#### Downgrade, then maybe upgrade

Adopting a custom element "downgrades" it to have a custom element state of "undefined." It then might become inserted into a document (perhaps immediately), which might cause an upgrade, if a definition with the matching name appears in the new document. That upgrade will necessarily proto-swizzle to the new definition's prototype, and run the new definition's constructor on the (already constructed oncee) object.

With this in place we could either:
- Leave the spec as-is for uncustomized elements (no proto swizzling).
- Change the spec to proto-swizzle uncustomized elements (including "downgraded" custom elements, which would proto-swizzle to the new document's HTMLElement, with subsequent upgrades potentially causing another proto-swizzle).

My understanding is that some browsers will not comply with either of these two options. And both seem consistent. So if we were to go this route I guess we'd probably leave the spec as-is for uncustomized elements out of inertia.

I am not sure introducing an adoptedCallback makes sense in this world, as it's unclear whose responsibility the adoption would be: the original definition, or the new definition (or both!?). In general the idea of adoptedCallback as a way of accomplishing something similar to built-in elements' "adoption steps" doesn't make sense here because in this world the two element definitions are largely unrelated, so it's not like it's the same type of "thing" and you just need to run some steps on some internal state. Indeed, after adoption the element will have two separate sets of internal state, one from each definition.

### Thoughts and conclusions

We have to pick one of these (or something new nobody thought of; let me know). Right now the spec specifies "Downgrade, then maybe upgrade."

One might think that the simplest position is "Throw when adopting," but I am not sure. It seems like that has its own unique challenges which might make it worse.

My personal preference at this point in time is "Stay custom with same definition". It seems self-consistent, and matches the built-ins. It has an extensibility point for adoptedCallback in the future. It even has a rationalization built in for the noncompliant browsers that I think avoids any future compatibility constraints. Anyway, "Downgrade, then maybe upgrade" seems increasingly bad the more I look at it.

Do others find this convincing?

**Comment by annevk** [comment]
Where you talk about printing above I think you are confusing cloning and adoption.

The `adoptedCallback` could also be an opt-in to prevent throwing/downgrade and retain identity of sorts, maybe. If that is feasible I think I would prefer going that route and then downgrade seems like the safest approach.

**Comment by dominiccooney** [comment]
@domenic What is the built-in rationalization for noncompliant browsers? I did not understand that part.

**Comment by domenic** [comment]
> Where you talk about printing above I think you are confusing cloning and adoption.

I don't believe so. My understanding was Gecko creates a new document and clones all children into the new document, which necessarily involves adoption. More info welcome, of course.

> The adoptedCallback could also be an opt-in to prevent throwing/downgrade and retain identity of sorts, maybe. If that is feasible I think I would prefer going that route and then downgrade seems like the safest approach.

This is meant to be a fourth proposal? Can you expand on it in an equivalent amount of detail? How does this work?

> What is the built-in rationalization for noncompliant browsers? I did not understand that part.

I was referring to my above

> It also allows the non-compliant browsers to rationalize their behavior by saying that the default adoptedCallback equivalent for built-in elements does proto-swizzling, even if the spec does not mention that.

**Comment by annevk** [comment]
(I would expect Gecko to simply clone the document.)

The Fourth Proposal(tm) is that by default custom elements downgrade, then maybe upgrade, per your definitions. However, if custom elements have a `adoptedCallback` (to be introduced at a later stage), they would not be downgraded (or proto-swizzled) and remain the same object.

Any proto-swizzling for nodes remains its own issue (except that with the `adoptedCallback` code path it would be up to the author-defined custom element to do that kind of thing).

So overall it's no different from the "Downgrade, then maybe upgrade" proposal, except that the future story is a little different (and better, imo).

**Comment by bzbarsky** [comment]
> Although only one built-in element, `<template>`, uses these.

Point of fact: `<img>` has nontrivial "adopting steps".

**Comment by domenic** [comment]
> The Fourth Proposal(tm) is that by default custom elements downgrade, then maybe upgrade, per your definitions. However, if custom elements have a adoptedCallback (to be introduced at a later stage), they would not be downgraded (or proto-swizzled) and remain the same object.

Which element definition is consulted for this opt-out? The source, I'd guess?

> So overall it's no different from the "Downgrade, then maybe upgrade" proposal, except that the future story is a little different (and better, imo).

And you're not bothered by the fact that it runs two different constructors on the same element by default? And doesn't match built-ins?

> Point of fact: <img> has nontrivial "adopting steps".

Thanks for the correction!

**Comment by dominiccooney** [comment]
One thing to be aware of if you go with adopting callbacks or adopting steps is keeping the custom element state in sync with anything that _reads_ or _sets_ the state.

For example, [Upgrade an element](https://html.spec.whatwg.org/multipage/scripting.html#upgrades) step 3 reads the state and step 10 sets the state. Author script may run between the time that reaction is queued and when it is run, or between those two steps (step 6.)

My impression is that reading or setting the custom element state outside of a reaction quickly leads to unfortunate edge cases. So if you do that, there should be some compensating action on the element's queue, like voiding its queued reactions. (The queue is reentrant, by the way, so you may have to look hard at the upgrade reaction and decide if it needs to abort.)

**Comment by bzbarsky** [comment]
> (I would expect Gecko to simply clone the document.)

For what it's worth, the clone Gecko performs for printing is a pretty special clone (e.g. when cloning a canvas we clone the actual image in the canvas, which is not the normal canvas cloning behavior).  We also don't really want to run script during this clone, I'm pretty sure.   And the clone is not exposed to script, so doesn't have detectable prototypes to start with.  So I don't think we should really block the discussion here on the clone-for-printing Gecko does.

**Comment by rniwa** [comment]
I think we should throw.  That's the easiest path for interoperability, and we can easily change that in the future to any of the other options but we can't do that once we picked any other _wrong_ option.

**Comment by domenic** [comment]
> I think we should throw. That's the easiest path for interoperability, and we can easily change that in the future to any of the other options but we can't do that once we picked any other wrong option.

Do you see a path toward us getting more information in the future, which we don't have now, that allows us to converge on another option? If we have all the information now, punting by choosing throw seems wrong.

---

I spoke to @esprehn and @ajklein about whether throwing is viable from an implementation perspective at BlinkOn. It seems like it is viable for us, however there is a potential issue. During the adoption agency algorithm (which does not _usually_ perform adoption, despite the name) there are a number of element removals and appends. I can't tell for sure but it seems possible you might be able to remove an element and re-insert it into another iframe.

I guess the thing to do in that case would be to add a "catch and report an exception" step to all insertion operations in the parser... not sure if that's OK.

I guess in general we have to worry about the fact that if adopt can throw, any insertion-based operation (insert, append, replace) can also throw now.

**Comment by rniwa** [comment]
> Do you see a path toward us getting more information in the future, which we don't have now, that allows us to converge on another option? If we have all the information now, punting by choosing throw seems wrong.

I don't think it's a matter of information as much as a matter of consensus.  We can't reach a consensus on custom elements behavior before reaching a consensus on builtin elements.

> the adoption agency algorithm

We investigated the adoption agency algorithm in the past and found that it can't create a custom element although it's possible to have `is=` on them.  This was yet another reason we didn't want to support `is=` since it's just not possible to run arbitrary scripts during the adoption agency algorithm safely.

> I guess in general we have to worry about the fact that if adopt can throw, any insertion-based operation (insert, append, replace) can also throw now.

This is generally true already because [**pre-insert**](https://dom.spec.whatwg.org/#concept-node-pre-insert) needs to [**ensure pre-insertion validity**](https://dom.spec.whatwg.org/#concept-node-ensure-pre-insertion-validity) which can throw a bunch of exceptions such as `HierarchyRequestError` and `NotFoundError`.

**Comment by domenic** [comment]
> We investigated the adoption agency algorithm in the past and found that it can't create a custom element although it's possible to have is= on them. This was yet another reason we didn't want to support is= since it's just not possible to run arbitrary scripts during the adoption agency algorithm safely.

By "create" do you mean "remove and re-insert"? I am not concerned about creating as much as with removing from one iframe and reinserting in another.

> This is generally true already because pre-insert needs to ensure pre-insertion validity which can throw a bunch of exceptions such as HierarchyRequestError and NotFoundError.

Right, but those exceptions cannot happen when called from browser code, because browser code never violates the invariants they are enforcing.

**Comment by domenic** [comment]
@annevk ping on my questions in https://github.com/w3c/webcomponents/issues/512#issuecomment-226415856

@rniwa ping on my above questions in https://github.com/w3c/webcomponents/issues/512#issuecomment-227309377.

I am still a bit frustrated by your insistence that

> We can't reach a consensus on custom elements behavior before reaching a consensus on builtin elements.

These issues can be separated. In this very thread, Gecko is interested in coming to consensus on custom elements, despite their unwillingness to follow the spec for built-in elements. (@annevk says, "Any proto-swizzling for nodes remains its own issue".) I don't think it's a good idea to block progress in custom elements based on this.

I think throwing creates an unnecessary and surprising inconsistency for authors, and wish you would be willing to entertain one of the other possibilities discussed in this thread, despite the fact that this is a historically tricky issue. We would like to provide authors some solution here.

**Comment by annevk** [comment]
> Which element definition is consulted for this opt-out? The source, I'd guess?

Yes.

> And you're not bothered by the fact that it runs two different constructors on the same element by default? And doesn't match built-ins?

My assumption is that the typical "by default" is that you won't find a suitable definition in the other global. That it might upgrade is mostly a side effect of it being a custom element. The way to match built-ins is the `adoptedCallback` extension that allows for preserving identity and state, as well as any proto-swizzling may we reach agreement on that.

All things I considered I think I prefer this over throwing, simply since throwing is such an unexpected behavior for these kind of operations.

**Comment by domenic** [comment]
> All things I considered I think I prefer this over throwing, simply since throwing is such an unexpected behavior for these kind of operations.

I also prefer it over throwing :). Can you say why you prefer it over "stay custom with same definition"?

**Comment by rniwa** [comment]
We would object to the whole notion of "downgrade" and calling constructors on the same element twice.  That's just unwarranted complexity we don't want in API nor in our engine.

**Comment by annevk** [comment]
I prefer it over retaining identity by default as that likely leads to leaking a global. Seems nicer if the default behavior does not do that. Having said that, Mozilla has said to be willing to live with that for custom elements before, so I won’t object.

**Comment by rniwa** [comment]
@annevk: you prefer what (it is ambiguous here)?

**Comment by annevk** [comment]
"Stay custom with same definition", I meant to reply to @domenic.

**Comment by rniwa** [comment]
Okay, so you're preferring to re-upgrading the element in the new document?

In that case, I would re-iterate our opposition for any notion of downgrades or re-upgrades in a new document.

**Comment by annevk** [comment]
Well, that is not all I said, right? I would prefer that to avoid leaking globals for custom elements. But I also mentioned that Mozilla had previously been okay with leaking globals for custom elements. I don't think that has changed.

**Comment by rniwa** [comment]
Sure. In that case, keeping the identity, that is not to make changes to the [[Prototype]] internal slot of the element adopted to a new document, and calling `adoptedCallback` on the custom element appears to be the path of least resistance.

**Comment by domenic** [comment]
Right, OK, it seems like we are converging on "Stay custom with same definition", with it not being Mozilla's preference, but they're willing to live with it. Thank you for being willing to compromise!

I'll work on speccing this over the next day or two, unless I misread and we're still not agreed.

Do people want to introduce adoptedCallback at the same time? Seems kind of nice, "while we're there". But on the other hand it also seems kind of silly to introduce adoptedCallback when all the other v2 callback ideas are still not discussed, including e.g. the much more common clonedCallback. I could go either way.

**Comment by rniwa** [comment]
I think we should introduce the callback given it would be only way for authors to avoid leaking the global of the old document.

**Comment by domenic** [comment]
Pull requests up at https://github.com/whatwg/dom/pull/279 and https://github.com/whatwg/html/pull/1510. Reviews and sign-off much appreciated.

Thank you all for working together on this!

**Comment by bzbarsky** [comment]
> Mozilla has said to be willing to live with that for custom elements before

More precisely, Mozilla has said it's willing to live with a way to opt in to the leaking behavior.  We're not super happy about it being the _default_ one.

**Comment by bzbarsky** [comment]
I guess the question is whether, if we're introducing the callback anyway, we could do something where the default behavior is non-leaky but the callback lets you opt in to the other behavior...

**Comment by rniwa** [comment]
> I guess the question is whether, if we're introducing the callback anyway, we could do something where the default behavior is non-leaky but the callback lets you opt in to the other behavior...

Given we're objecting to downgrades, what kind of a behavior do you suppose will accomplish what you want here and has a chance of reaching a consensus?

**Comment by bzbarsky** [comment]
I don't know.  If I had a concrete proposal, I would have made it...

**Comment by annevk** [comment]
The only alternative I see is that you throw by default and opting into the callback makes you preserve identity and not throw. However, that will just lead to no-op callbacks and frustration about custom elements not moving around as normal elements do.

**Comment by domenic** [comment]
OK, @dominiccooney pointed out a problem with "stay custom with same definition" semantics, and now I am not sure we are on the right path here.

### The problem

A common pattern is for people to use a `<template>` to instantiate their element's contents. However, this pattern fails with our current setup for custom element registries and our proposed setup for adopting. Consider:

``` html
<!DOCTYPE html>
<template id="xfoo">
  <x-bar>hello</x-bar>
</template>

<script>
"use strict";
customElements.define("x-bar", class XBar extends HTMLElement { ... });

const template = document.querySelector("template#xfoo");

class XFoo extends HTMLElement {
  constructor() {
    super();
    this.attachShadow({ mode: "open" });
    this.shadowRoot.appendChild(template.content.cloneNode(true));
  }
}

customElements.define("x-foo", XFoo);
</script>
```

The problem is, template contents belong to a special browsing context-less Document. Thus, any elements created inside the template contents are never custom (step 2 of [look up a custom element definition](https://html.spec.whatwg.org/multipage/scripting.html#look-up-a-custom-element-definition) returns null). Cloning the template contents doesn't change that. And then, per the proposed adoption semantics in this thread, when we append the resulting document fragment, the cloned non-custom elements will carry their non-customness with them into XFoo's shadowRoot---not at all what we intended.

I hope we can all agree this is a problem? This kind of template + custom elements + shadow DOM scenario has been part of the web components story for as long as I have been involved, and it seems like we accidentally broke it. (Or at least, discovered that it wasn't on very solid footing.)

### Proposed solutions

@dominiccooney and I came up with several solutions, trying to take into account the various opinions expressed in this thread about defaults, added concepts, and more.
- "Auto downgrade": this is essentially the old "downgrade, then upgrade" from https://github.com/w3c/webcomponents/issues/512#issuecomment-226281431. This has the drawback of running a constructor twice. However, `adoptedCallback()` has the potential of allowing some kind of teardown/reset of state, which could help...
- "Manual downgrade": we keep the "stay custom with same definition" semantics by default, but elements which wish to work with `<template>` can use some new API, say `customElements.downgrade(element)`, inside their `adoptedCallback()`.
- "Branching adoption": when adopting elements, if the element already has a custom element definition, keep it. However, if it does not have one, then look up the potential custom element definition in the destination document, and if such a definition exists, assign it. This prevents running any constructors twice, and solves the template problem (as well as the similar-but-less-common `responseXML` and `DOMParser` problems).
  - There's a variant of this where we branch based on whether the source document had a browsing context, instead of based on whether the being-adopted element had a custom element definition. The difference being that "has a browsing context" would only catch the template/responseXML/DOMParser cases, whereas "has a custom element definition" would also allow moving elements between "normal" documents and the element "lighting up" if only the destination document has a definition.

Of these I am warming up to "branching adoption"... The "has a browsing context" version even seems somewhat principled. I guess we conceptualize that as "elements get a chance to become custom the first time their node document has a browsing context", which is either at creation, or at adoption.

I'm really sorry to retread this after we already went through the discussion and seemingly settled on a good solution. But I think we missed something big, and fixing it will be worth it!

**Comment by ajklein** [comment]
Changing the example to use `importNode` should work as expected, unless I'm missing something:

``` js
this.shadowRoot.appendChild(document.importNode(template.content, true));
```

**Comment by domenic** [comment]
Yeah... my mega-post may have been based on a misunderstanding... thanks @ajklein -_-. I'll wait for @dominiccooney to confirm, but I think this may have been a false alarm...

**Comment by dominiccooney** [comment]
Sorry for the spam about importNode everyone; importNode is fine. Thank you @ajklein  for pointing out my misapprehension. Hope this did not derail the important discussion about adoption!

**Comment by annevk** [comment]
FYI: I plan on reviewing and working with @domenic on landing https://github.com/whatwg/dom/pull/279 and https://github.com/whatwg/html/pull/1510 this week. That will close this issue per https://github.com/w3c/webcomponents/issues/512#issuecomment-230764837. The disagreement from (some at) Mozilla is noted, but unfortunately there doesn't seem to be a reasonable alternative.
